crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -1891,19 +1891,18 @@ where
                     flake8_bugbear::rules::f_string_docstring(self, body);
                 }
                 let definition = docstrings::extraction::extract(
-                    &self.ctx.visible_scope,
+                    self.ctx.visible_scope,
                     stmt,
                     body,
-                    &Documentable::Function,
+                    Documentable::Function,
                 );
                 if self.settings.rules.enabled(Rule::YieldInForLoop) {
                     pyupgrade::rules::yield_in_for_loop(self, stmt);
                 }
-                let scope =
-                    transition_scope(&self.ctx.visible_scope, stmt, &Documentable::Function);
+                let scope = transition_scope(self.ctx.visible_scope, stmt, Documentable::Function);
                 self.deferred.definitions.push((
                     definition,
-                    scope.visibility.clone(),
+                    scope.visibility,
                     (self.ctx.scope_stack.clone(), self.ctx.parents.clone()),
```

Comment:
Cloning the whole stack here (and when adding to `dead_scopes` is something we should look at at some point. 

I think we could reduce the cost of cloning by:

* Having one `Vec<ScopeInternal>` that stores all existing scopes
* `ScopeInternal` is a `Scope` without its `id`. The is implicit by the position of the `Scope` in the vector. 
* Add a `parent: Option<ScopeId>` to scope to allow upwards traversal. 
* Change `ScopeStack` to `Vec<ScopeId>` 

I need to look closer into what `parents` is used for

Reply:
`parents` is the stack of statements that are parents of the current statement (so e.g. if we're looking at a function definition inside of a class, it might have a `parents` of `vec![ClassDef { ... }]`).